### ISSUE TITLE
Indirect array accesses not ZP-optimized

### ISSUE THREAD

**Issue by clbr** [cc65/cc65#243]
It seems accesses of the type "array[var]" do not get the one-byte (or more) reduction, even when both array and var are in ZP. The direct type, "array[13]", does get it.

All under -Oisr.

**Comment by mrdudz** [comment]
can you be more specific? i think what you want the compiler to do is actually not possible, as it doesnt know how large the offset in "array[var]" is going to be, and thus (correctly) generates code that can access all 64k.

**Comment by clbr** [comment]
It's not possible for it to wrap out of ZP without using asm or linker tricks, so the compiler is safe in assuming the access stays in ZP.

```
#pragma bss-name (push, "ZEROPAGE")
u8 arr[257];
#pragma bss-name (pop)
```

This doesn't compile. The linker will bail since the array exceeds the zeropage size. Only by using linker tricks to place symbols at both end of ZP and beginning of RAM could you get defined behavior by going out of bounds, it is safe to assume nobody will do such a pointless thing.

**Comment by clbr** [comment]
Oh, the indexing variable is of course an u8, and usually used in a loop of the type
for (i = 0; i < 16; i++)

A more advanced compiler can thus prove it will never overflow. A simpler one can rely on the linker as mentioned in my post above.

**Comment by mrdudz** [comment]
> It's not possible for it to wrap out of ZP without using asm or linker tricks, so the compiler is safe in assuming the access stays in ZP.

actually, thats not quite true - it can always be accessed using indirect adressing (which is what is normally used for pointers and arrays). the linker problem you mention is a different thing and not (directly) related to that. i'm not sure if relying on the linker behaviour is a good idea at this point.

**Comment by clbr** [comment]
> actually, thats not quite true - it can always be accessed using indirect adressing (which is what is normally used for pointers and arrays). the linker problem you mention is a different thing and not (directly) related to that. i'm not sure if relying on the linker behaviour is a good idea at this point.

That's still undefined buffer overflow, no?

// ZP pragmas around
u8 array[16];
u8 i;
u8 *ptr;
// end ZP pragma

i = 32;
array[i]; // undefined
ptr = &array[15];
ptr++;
*ptr; // undefined

All arrays in there have to fit in the section, and buffer overflow is
undefined.

**Comment by mrdudz** [comment]
ptr = &array[15]; 
ptr++; 
*ptr; // undefined 

this is not undefined - it will access the byte behind that array :)

**Comment by acqn** [comment]
Is the original point of this issue that the compiler currently can't be told and take benefit on that a variable defined in the C source is going to be put on ZP in the link stage? If so this issue is related to #917 very much.

I think the compiler can just let the programmer specify what to be put on ZP (like in my solution for #917) and be fully responsible for writing the correct code. In theory the compiler may be able to automatically put some variables on ZP with any annotations, but that is a bit too advanced....

**Comment by clbr** [comment]
The compiler knows the array and the index are in ZP, but it doesn't use the ZP,x format which would save one byte in size.

**Comment by mrdudz** [comment]
> I think the compiler can just let the programmer specify what to be put on ZP (like in my solution for #917) and be fully responsible for writing the correct code.

yes. it should never do this automatically - which will create various problems

about the original issue, i am still not sure it can be solved in a way that works generically, for all possible cases
